Makefile

```diff
@@ -1,23 +1,137 @@
-hello:
-	@echo "Please inspect Makefile for list of commands"
+.PHONY: clean black requirements docs clean unit_test_data unit_test unit_test_wip unit_test_entrypoints
 
+#################################################################################
+# GLOBALS                                                                       #
+#################################################################################
+
+PROJECT_DIR := $(shell dirname $(realpath $(lastword $(MAKEFILE_LIST))))
+PROJECT_NAME = autometa
+PYTHON_INTERPRETER = python3
+
+ifeq (,$(shell which conda))
+HAS_CONDA=False
+else
+HAS_CONDA=True
+endif
+
+#################################################################################
+# COMMANDS                                                                      #
+#################################################################################
+
+## Delete all compiled Python files
+clean:
+	rm -rf htmlcov && make clean -C docs
+	find . -type f -name "*.py[co]" -delete
+	find . -type d -name "__pycache__" -delete
+
+## Apply black formatting
+black:
+	black --exclude autometa/validation autometa
+
+## Set up python interpreter environment
+create_environment:
+ifeq (True,$(HAS_CONDA))
+		@echo ">>> Detected conda, creating conda environment."
+ifeq (3,$(findstring 3,$(PYTHON_INTERPRETER)))
+	conda create --name $(PROJECT_NAME) python=3
+else
+	conda create --name $(PROJECT_NAME) python=2.7
+endif
+	@echo ">>> New conda env created. Activate with:\nsource activate $(PROJECT_NAME)"
+else
+	$(PYTHON_INTERPRETER) -m pip install -q virtualenv virtualenvwrapper
+	@echo ">>> Installing virtualenvwrapper if not already installed.\nMake sure the following lines are in shell startup file\n\
+	export WORKON_HOME=$$HOME/.virtualenvs\nexport PROJECT_HOME=$$HOME/Devel\nsource /usr/local/bin/virtualenvwrapper.sh\n"
+	@bash -c "source `which virtualenvwrapper.sh`;mkvirtualenv $(PROJECT_NAME) --python=$(PYTHON_INTERPRETER)"
+	@echo ">>> New virtualenv created. Activate with:\nworkon $(PROJECT_NAME)"
+endif
+
+#################################################################################
+# PROJECT RULES                                                                 #
+#################################################################################
+
+## Install autometa from source
 install:
-	python setup.py install
+	$(PYTHON_INTERPRETER) setup.py install
 
+## Build documentation for autometa.readthedocs.io
 docs:
 	make clean html -C docs && open docs/build/html/index.html
 
-clean:
-	@echo "Removing everything under 'htmlcov'..."
-	@rm -rf htmlcov && make clean -C docs
+## Build test_data.json file for unit testing
+unit_test_data: tests/data/
+	$(PYTHON_INTERPRETER) make_test_data.py
```

Comment:
I'm not sure it would be necessary to have make available in the docker image, since this is not used at any stage during the pipeline execution

Reply:
Yes, but that is how I am testing the code in a clean environment.

Reply:
I'm not sure I understand

Reply:
I would like to have the ability to run the unit tests in a Docker so I can be sure that the environment is right. This was actually in the context of checking if the documentation is complete in its installation procedure.

Reply:
> BTW I did install `build-essential` in the docker image so that `make` is available.

`make` can be installed via conda and has been added to `requirements.txt`, so this is immediately available without having to use `build-essential`